### ISSUE TITLE
Add "raw" endpoint to return raw, unparsed request data?

### ISSUE THREAD

**Issue by llamasoft** [kennethreitz/httpbin#423]
If possible, please consider adding a `/raw` or `/echo` endpoint that returns the raw HTTP request that the server received.  
This feature can be used to help diagnose misbehaving applications (e.g. sending duplicate headers with different values or using incorrect line endings) or debug applications using niche HTTP features (e.g. sub-headers for `multipart/form-data` sections).

**Comment by sigmavirus24** [comment]
Before HTTPbin the application even sees the request it is parsed. We cannot possibly return the raw unparsed request data.

Thanks for the suggestion, new ideas are always welcome.

**Comment by llamasoft** [comment]
I did some tinkering and managed to get the raw request body returned correctly.  The trick is using [`request.get_data()`](https://stackoverflow.com/a/23898949/477563) before accessing the `request.data`, `request.json`, or `request.form` members.  Until those members are accessed, the request body is cached and remains unparsed.  After accessing one of those members, the data is parsed and purged.  
Here is a small working example I came up with:

```
@app.route('/body', methods=['GET', 'POST', 'PUT', 'DELETE', 'PATCH', 'TRACE'])
def view_body():
    """Simple Html Page"""
    raw_body = request.get_data()
    response = get_dict('url', 'args', 'headers', 'origin', 'method')
    response['body'] = raw_body
    return jsonify(response)

@app.route('/body/raw', methods=['GET', 'POST', 'PUT', 'DELETE', 'PATCH', 'TRACE'])
def view_raw_body():
    """Simple Html Page"""
    return request.get_data()
```

As for headers, I haven't had much luck.  Allegedly, [`request.headers`](http://werkzeug.pocoo.org/docs/0.14/datastructures/#werkzeug.datastructures.Headers) are supposed to be a "dict-like interface but is ordered and can store the same keys multiple times" but somewhere along the line something (gunicorn, werkzeug, or Flask) ends up butchering the ordering and merging duplicate headers into comma-delimited values.  Still, having the raw request body available should be useful for debugging.
